### ISSUE TITLE
Failure to compile DLL

### ISSUE THREAD

**Issue by marius-sucan** [brechtsanders/winlibs_mingw#115]
Hello!

I am trying to compile a x64 DLL on Windows 10. I tried with these options to compile: gcc -c -mno-cygwin mydll.cpp , but then it suggested to use -mno-clwb  ...

So, I used
gcc -c -mno-clwb mydll.cpp

And after that:
gcc -shared -o mydll.dll mydll.o -Wl,--kill-at,--out-implib,libmydll.a

But then I get an entire screen of things like this:
f:/temp/winlibs-x86_64-gcc-12.1.0-mingw-w64msvcrt-10.0.0-r3/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: testus.o:testus.cpp:(.text$_ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPcEEvT_S7_St20forward_iterator_tag[_ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPcEEvT_S7_St20forward_iterator_tag]+0xac): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_S_copy_chars(char*, char*, char*)'
f:/temp/winlibs-x86_64-gcc-12.1.0-mingw-w64msvcrt-10.0.0-r3/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: testus.o:testus.cpp:(.text$_ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPcEEvT_S7_St20forward_iterator_tag[_ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPcEEvT_S7_St20forward_iterator_tag]+0xc4): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_set_length(unsigned long long)'
f:/temp/winlibs-x86_64-gcc-12.1.0-mingw-w64msvcrt-10.0.0-r3/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: testus.o:testus.cpp:(.text$_ZNSt16allocator_traitsISaIcEE37select_on_container_copy_constructionERKS0_[_ZNSt16allocator_traitsISaIcEE37select_on_container_copy_constructionERKS0_]+0x1c): undefined reference to `std::allocator<char>::allocator(std::allocator<char> const&)'

In other words, many undefined references.

What am I doing wrong?

In the header file I have this:
#define DLL_API extern "C" __declspec(dllexport)
#define DLL_CALLCONV __stdcall

And the functions are defined like this:
DLL_API int DLL_CALLCONV myFuncu(const char* a, int b) {
 ...............
}

Thank you.
Best regards, Marius.

**Comment by brechtsanders** [comment]
If it's C++ make sure to use `g++` instead of `gcc`

**Comment by marius-sucan** [comment]
Thank you very much .

Based on your suggestion I used this:

g++ -c mydll.cpp
g++ -shared -o mydll.dll mydll.o -W

The DLL is created, but... when I use "LoadLibrary" I get error 126 on it. I do not know why.

When I inspect the DLL, the functions I defined are public, with DLL Export Viewer from Nirsoft.

I do not know why I have issues with LoadLibrary. I apologize for asking ... do you have any ideas why this could be the case?

Best regards, Marius.

**Comment by brechtsanders** [comment]
You should check the DLL with [Dependency Walker](https://www.dependencywalker.com/).
Things to check there are if all dependency DLLs can be found and loaded and if there is no mix between 32-bit and 64-bit.
Since this is not an actual issue I will close this ticket. These kinds of questions are better asked on a forum like https://stackoverflow.com/

**Comment by marius-sucan** [comment]
Hello. My apologies for coming back to this, but...

Based on the Dependecy Walker application, the compiled DLL is dependent on many API-MS-WIN-*.DLL files which cannot be found on my system. As previously stated, I run on Windows 10, and I do have such files on my system, but not the versions the G++ compiled DLL is dependent on.  It is exactly what I want to avoid. This is the very reason I do not want to compile my DLL with MS VS 2019. I do not want my DLL to be dependent on these set of DLLs.

What version of WinLib to use to this end? Or... what are the correct command line options to have my DLL compiled without relying on any API-MS-WIN-*.DLL file?

Thank you . Best regards, Marius.

**Comment by brechtsanders** [comment]
You can safely ignore the dependency warnings `API-MS-WIN-*.DLL`. What is important that no other dependency files are missing and no errors are reported by Dependency Walker.

I don't understand why you were using `LoadLibrary()`.

Do you actually have problems using your library from an application that calls it by linking to it?

**Comment by marius-sucan** [comment]
Yes, I do have to link to it via LoadLibrary().  I got it working on Windows 10 by adding to g++ these options:   -static-libgcc -static-libstdc++ . But, on Windows 7 it still does not work.

Thank you for the reply.

Best regards, Marius.

**Comment by marius-sucan** [comment]
For future reference, if anyone has this issue...

On Windows 7 ... libwinpthread-1.dll is required.

Best regards, Marius.

**Comment by brechtsanders** [comment]
That's not related to the Windows version. If your application (or DLL) depends on it (and it wasn't build statically) you will need to distribute it. Maybe you already had it somewhere in your PATH on the systems where it worked...

I write a tool to list/copy dependancies that may be useful: https://github.com/brechtsanders/pedeps

**Comment by 3oheicrw** [comment]
> Based on the Dependency Walker application

Dependency Walker is no longer developed. Use this instead:

https://github.com/lucasg/Dependencies
